src/useFormState.ts

```diff
@@ -128,6 +137,80 @@ export function useFormState<T, I>(opts: UseFormStateOpts<T, I>): ObjectState<T>
   return form;
 }
 
+export type ObjectStateCache<T> = Record<string, [ObjectState<T>, T]>;
+type UseFormStatesOpts<T> = {
+  config: ObjectConfig<T>;
+  autoSave?: (state: ObjectState<T>) => Promise<void>;
+  getId: (t: T) => string;
+};
+type GetObjectStateProps<T, I> = {
+  readOnly?: boolean;
+  init?:
+    | T
+    | {
+        input: I;
+        map: (input: Exclude<I, null | undefined>) => T;
+      };
+  addRules?: (state: ObjectState<T>) => void;
```

Comment:
I'd probably do this as a `useFormStates` prop b/c the logic should be similar for each row.

Same with `map`, i.e. I think the only input to `getObjectState` might just be `I`.

```
const formState = getObjectState(someFragment);
```

I believe is what they're doing in schedules v2.